capstone/src/test/java/com/google/sps/servlets/LoginServletTest.java

```diff
@@ -0,0 +1,97 @@
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import java.util.HashMap;
+
+/**
+ * Unit tests for {@link LoginServlet}.
+ */
+ @RunWith(JUnit4.class)
+public class LoginServletTest {
+  private static final String USER_EMAIL = "test@mctest.com";
+  private static final String USER_ID = "testy-mc-test";
+
+  // Set no eventual consistency, that way queries return all results.
+  // https://cloud.google.com/appengine/docs/java/tools/localunittesting
+  // #Java_Writing_High_Replication_Datastore_tests
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+            new LocalDatastoreServiceTestConfig()
+                .setDefaultHighRepJobPolicyUnappliedJobPercentage(0),
+            new LocalUserServiceTestConfig())
+          .setEnvEmail(USER_EMAIL)
+          .setEnvIsLoggedIn(true)
+          .setEnvAuthDomain("gmail.com")
+          .setEnvAttributes(
+              new HashMap(
+                  ImmutableMap.of(
+                      "com.google.appengine.api.users.UserService.user_id_key", USER_ID)));
+
+  @Mock private HttpServletRequest mockRequest;
+  @Mock private HttpServletResponse mockResponse;
+  private StringWriter responseWriter;
+  private DatastoreService datastore;
+  private LoginServlet loginServlet;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Set up a fake HTTP response.
+    responseWriter = new StringWriter();
+    when(mockResponse.getWriter()).thenReturn(new PrintWriter(responseWriter));
+
+    loginServlet = new LoginServlet();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGet_userIsLoggedIn() throws Exception {
+    loginServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+    assertThat(response).contains(USER_EMAIL);
+  }
+
+  @Test
+  public void doGet_userIsLoggedOut() throws Exception {
+    helper.setEnvIsLoggedIn(false);
+
+    loginServlet.doGet(mockRequest, mockResponse);
+    String response = responseWriter.toString();
+
+    assertThat(response).contains("false");
+    assertFalse(response.contains(USER_EMAIL));
+
+    helper.setEnvIsLoggedIn(true);
```

Comment:
Is this line necessary? You seem to be mocking state after the test, and not testing it at all. And if you were to add additional assertions after this line, would you be testing the servlet, or the helper at that point?

Reply:
If you're trying to restore state after the test, I think this should already be handled in `helper.setUp()` and `helper.tearDown()`

and if not (please confirm), then I would DEFINITELY move this line into `setUp` because attempting to restore state in a test is something which is not guaranteed. Assume a test can crash and fail at any point, whereas you can safely assume setUp is guaranteed to run before a test runs.

Reply:
Okay, yeah, I was trying to restore it's state after the test. I didn't realize that setUp() and tearDown() were run before and after each test